### ISSUE TITLE
Memory leak in conv2d()

### ISSUE THREAD

**Issue by w4nderlust** [tensorflow/tensorflow#14171]
### System information

Python: 3.5.2
TensorFlow version 1.3.0 via pip, 1.3.1 via source, 1.4.0-rc1 from source (confirmed the problem on all three of them)
Bazel: 0.7.0
Ubuntu 16.04
GPU: GeForce GTX 1080 with 8GB
Nvidia drivers: 384.90
CUDA: 8.0
CuDNN: 6.0
Optimizer: Adam with default parameters

### Problem Description

I will post here some evidence that makes me think that there's a major leak in `conv_2d()`.
Unfortunately I can't provide code and data replicate this as my employer doesn't allow me to as both are confidential, but hopefully the information i will provide will make it possible for you to replicate it.

The setting:
I'm training a word cnn on text data on GPU. The length of the sequence of text is 256, the dimension of the embeddings is 256, I have 4 convolutional layers in parallel with different sizes (2, 3, 4 and 5 x 256). The inputs to the conv layers are `batch_size x 256 x 256 x 1`.
In my code I run a full epoch of training, then I evaluate on the training set, the validation set and the test set, then I start a new epoch.

The problem:
Depending on the batch_size the memory needed by TensorFlow to compute his operations changes, but it is in the order of few hundreds MB. At the last batch of training, the memory consumption increases dramatically, as happens at the last batch of evaluation on validation set and the last batch of evaluation on the test set.
As _traini_set_size mod batch_size is not 0 (the same happens for validation and test set too), the last batch has a different dimension with respect to all others batches. For instance in my case with a batch_size of 100, the last batch of the validation set is of size 25.
What happens is that, in the `conv2d()` function TF allocates a lot of memory if the batch size is not the same that was used so far. In my case, that operation goes from needing 50 MB to 3.18 GB. As it seems too much, I suspect there's a memory leak somehow.

I'm attaching 2 screenshots taken from TensorBoard. What they show is the same node in the second last batch of the validation set and the last batch of the validation set. The memory consumption is in the node state on the right of the image. I can share the log directory if needed as it contains the graph of the model and the memory information at all steps of training and evaluation.

![eval_vali_step_8](https://user-images.githubusercontent.com/349256/32309260-99a53ab4-bf47-11e7-96d0-a429180f0e1c.png)
![eval_vali_step_9](https://user-images.githubusercontent.com/349256/32309259-99895934-bf47-11e7-8b3f-08306321cb8e.png)

The weird thing is that when batch_size is 256 or 512 it doesn't happen, but with batch_size 100 or 128 or 200 (the other 3 I tested) it happens. testing on another bigger dataset, even with 256 as the batch size the supposed memory leak happens, but in that case TF tries to allocate around 33 GB of ram and goes out of memory. My temporary workaround is just throwing away the last batch of the train, validation and test set. Doing so the memory consumption keeps constant.

Hopefully this is enough information for investigate the problem, otherwise feel free to request me additional information, even if, as I said, unfortunately I can't provide code and data.

**Comment by jart** [comment]
If you found a leak in conv2d, that would be a very high impact find, because it's such a commonly used feature. While it's less likely there would be a major issue, such a report is always worthy of consideration.

Is it possible that your shape, filter, stride, sizeof(dtype), etc. are all sort of multiplying together into that huge amount of memory? I've personally always found it to be interesting how quickly tiny integer parameters can turn into all the RAM that exists on my system. I wouldn't be surprised if the same thing has happened to every one of my colleagues.

If you think that's unlikely to be the case, then we would probably need a minimal reproducible example in order to investigate further.

**Comment by w4nderlust** [comment]
Hey jart, thanks for you answer. (I use you fabulous lib, it’s amazing!)

I spent 4 days debugging this issue in my code and I ruled out all other possibilities. I looked at the datapoints in the last batches one by one and also reshuffled all the datasets to be sure it was not a data related issue. Unfortunately no, it doesn’t seem to me that it is an issue related with loosing track of the sizes of the tensors, also because the last batch is actually smaller than the other batches, so the operation should actually use less memory and not more.

I realize that con2d is second only to add and matmul in terms of usage, so if no one encountered this issue before it could be that it happens to me because of a weird combination of os+python+gpu+nvidia software version, but I believe it’s worth investigating nonetheless.

I can try to extract a subset of my codebase and generate some synthetic data with the same dimensions of the offending dataset, but it will require me some time. I wouldn’t be able to do it before the weekend.

**Comment by jart** [comment]
If you liked Fabulous, check out [hiptext](https://github.com/jart/hiptext) which is what came next. It can [render Gangnam Style](https://www.youtube.com/watch?v=sYMczHySmYU) with quasi 512-color unicode half blocks in Terminal.app. @keroserene and I used it to make [rickrollrc](https://github.com/keroserene/rickrollrc). @tombh [plugged hiptext into X11](https://github.com/tombh/texttop) so you can use Google Chrome in your terminal.

I trust your judgement and appreciate the time you've invested so far. We'll be keeping this one open until at least the weekend. It's worth noting that at Google we have policies for situations like this, where if it's under a certain number of lines, it's less of a concern—especially if it helps the community. So a minimal reproducible example is likely to keep everyone happy. Just be sure to check with your employer.

**Comment by w4nderlust** [comment]
Here's a python3 script to reproduce the issue (it's a bit long but I tried to be as faithful to my original setting even if I'm not using a single line of code from my employer):

```python
import math

import numpy as np
import tensorflow as tf

# parameters
skip_last = False

train_set_size = 7113
vali_set_size = 925
test_set_size = 1962

text_length = 256
vocab_size = 20000
num_classes = 500

epochs = 2
batch_size = 128

# create artificial dataset
x_train = np.random.randint(0, vocab_size, [train_set_size, text_length])
y_train = np.random.randint(0, num_classes, [train_set_size])
x_vali = np.random.randint(0, vocab_size, [vali_set_size, text_length])
y_vali = np.random.randint(0, num_classes, [vali_set_size])
x_test = np.random.randint(0, vocab_size, [test_set_size, text_length])
y_test = np.random.randint(0, num_classes, [test_set_size])


# Building model
def fc_layer(inputs, in_count, out_count, act="relu"):
    weights = tf.get_variable("weights", [in_count, out_count])
    bias = tf.get_variable("bias", [out_count])
    layer = tf.matmul(inputs, weights) + bias
    if act:
        return getattr(tf.nn, act)(layer)
    else:
        return layer


def conv_layer(inputs, kernel_shape, bias_shape, stride=1, act="relu", padding='VALID', dropout=None, is_training=True):
    weights = tf.get_variable("weights", kernel_shape)
    conv = tf.nn.conv2d(inputs, weights,
                        strides=[1, stride, stride, 1], padding=padding)
    layer = conv
    if act:
        layer = getattr(tf.nn, act)(layer)
    if dropout is not None:
        layer = tf.layers.dropout(layer, rate=dropout, training=is_training)
    return layer


def word_cnn(input_text,
             vocab_size,
             dropout,
             embedding_size=256,
             filter_sizes=(2, 3, 4, 5),
             num_filters=(256, 256, 256, 256),
             fully_connected_sizes=(512, 256),
             regularize_layers=(False, True, True),
             embeddings_oon_cpu=False,
             is_training=True,
             name=None,
             **kwargs):
    assert len(filter_sizes) == len(num_filters)
    assert len(regularize_layers) == len(fully_connected_sizes) + 1

    sequence_length = input_text.get_shape()[-1]

    with tf.variable_scope("word_cnn_{}".format(name)):

        # ================ Word Embeddings ================
        if embeddings_oon_cpu:
            with tf.device('/cpu:0'):
                word_embeddings = tf.Variable(tf.random_uniform([vocab_size, embedding_size], -1.0, 1.0),
                                              name='word_embeddings')
                initial_input = tf.nn.embedding_lookup(word_embeddings, input_text, name='embeddings_lookup')
        else:
            word_embeddings = tf.Variable(tf.random_uniform([vocab_size, embedding_size], -1.0, 1.0),
                                          name='word_embeddings')
            initial_input = tf.nn.embedding_lookup(word_embeddings, input_text, name='embeddings_lookup')
        curr_input = tf.expand_dims(initial_input, -1)

        # ================ Conv Layers ================
        parallel_conv_layers = []
        for i in range(len(filter_sizes)):
            curr_filter_size = filter_sizes[i]
            curr_num_filters = num_filters[i]
            with tf.variable_scope("conv_{}".format(curr_filter_size)):
                filter_shape = [curr_filter_size, embedding_size, 1, curr_num_filters]
                layer_output = conv_layer(curr_input, filter_shape, [curr_num_filters],
                                          is_training=is_training, dropout=None)
                layer_output = tf.nn.max_pool(
                    layer_output,
                    ksize=[1, sequence_length - curr_filter_size + 1, 1, 1],
                    strides=[1, 1, 1, 1],
                    padding='VALID',
                    name="pool_{}".format(curr_filter_size))
                parallel_conv_layers.append(layer_output)

        # Flatten to vector
        num_filters_total = sum(num_filters)
        hidden = tf.reshape(tf.concat(parallel_conv_layers, 3), [-1, num_filters_total], name='flatten')
        hidden_size = num_filters_total

        # ================ Fully Connected Layers ================
        num_fully_connected_layers = len(fully_connected_sizes)
        for i in range(num_fully_connected_layers):
            fully_connected_size = fully_connected_sizes[i]
            with tf.variable_scope("fc_{}".format(i)):
                if dropout is not None:
                    with tf.name_scope("dropout"):
                        hidden = tf.layers.dropout(hidden, rate=dropout, training=is_training)

                # Fully Connected Layer
                hidden = fc_layer(hidden, hidden_size, fully_connected_size)
                hidden_size = fully_connected_size

    return hidden, hidden_size


is_training = tf.placeholder(tf.bool, [], name='is_training')
global_step = tf.Variable(0, trainable=False)
dropout = tf.placeholder(tf.float32, name="dropout")

input_placeholder = tf.placeholder(tf.int32, [None, text_length], name="input_placeholder")
label_placeholder = tf.placeholder(tf.int64, [None], name="label_placeholder")

hidden, hidden_size = word_cnn(input_placeholder, vocab_size, dropout=dropout, is_training=is_training, name="text")
softmax_w = tf.get_variable("softmax_weights", [hidden_size, num_classes])
softmax_b = tf.get_variable("softmax_bias", [num_classes])
logits = tf.matmul(hidden, softmax_w) + softmax_b

probabilities = tf.nn.softmax(logits, name="probabilities")
predictions = tf.argmax(logits, 1, name="predictions")
with tf.device('/cpu:0'):
    top_k_predictions = tf.nn.top_k(logits, k=3, sorted=True, name="top_k_predictions")

correct_predictions = tf.equal(predictions, label_placeholder, name="correct_predictions")
with tf.device('/cpu:0'):
    hits_at_k = tf.nn.in_top_k(logits, label_placeholder, 3, name="hits_at_k")

onehot_labels = tf.one_hot(label_placeholder, num_classes, name="onehot_labels")
loss = tf.losses.softmax_cross_entropy(onehot_labels=onehot_labels, logits=logits)

optimize = tf.train.AdamOptimizer(0.0025).minimize(loss)


# training
def convert_size(size_bytes):
    if size_bytes == 0:
        return "0B"
    size_name = ("B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
    i = int(math.floor(math.log(size_bytes, 1024)))
    p = math.pow(1024, i)
    s = round(size_bytes / p, 2)
    return "{} {}".format(s, size_name[i])


class Batcher(object):
    def __init__(self, dataset, batch_size, skip_last=False):
        self.dataset = dataset
        self.batch_size = batch_size
        self.total_size = min(map(len, dataset.values()))
        self.skip_last = skip_last
        self.index = 0

    def next_batch(self):
        if self.last_batch():
            self.index = 0
        sub_batch = {}
        for key in self.dataset:
            sub_batch[key] = self.dataset[key][self.index:self.index + self.batch_size]
        self.index += self.batch_size
        return sub_batch

    def last_batch(self):
        return self.index >= self.total_size or (self.skip_last and self.index + self.batch_size >= self.total_size)


def batch_eval(session, dataset, batch_size, epoch, name):
    batcher = Batcher(dataset, batch_size, skip_last)
    tot_size = batcher.total_size
    losses = []
    correct_preds = []
    hits_at_ks = []
    step = 0
    while not batcher.last_batch():
        batch = batcher.next_batch()
        loss_val, correct_predictions_val, hits_at_k_val = session.run(
            [loss, correct_predictions, hits_at_k],
            feed_dict={input_placeholder: batch['x'], label_placeholder: batch['y'], dropout: 0.3, is_training: True}
        )
        losses.append(loss_val)
        correct_preds.append(correct_predictions_val)
        hits_at_ks.append(hits_at_k_val)
        print('epoch', epoch, name, 'eval step', step)
        mbiu = session.run(tf.contrib.memory_stats.MaxBytesInUse())
        print('mbiu', convert_size(mbiu))
        step += 1
    losses = np.array(losses)
    correct_preds = np.concatenate(correct_preds)
    hits_at_ks = np.concatenate(hits_at_ks)
    return losses.sum() / tot_size, correct_preds.sum() / tot_size, hits_at_ks.sum() / tot_size


with tf.Session() as session:
    tf.global_variables_initializer().run()

    epoch = 0
    step = 0
    train_batcher = Batcher({'x': x_train, 'y': y_train}, batch_size, skip_last)

    while epoch < epochs:
        if step == 0:
            print("Epoch {}".format(epoch + 1))
        batch = train_batcher.next_batch()
        _, loss_val = session.run(
            [optimize, loss],
            feed_dict={input_placeholder: batch['x'], label_placeholder: batch['y'], dropout: 0.3, is_training: True}
        )
        print('epoch', epoch + 1, 'step', step + 1, 'loss', loss_val)
        mbiu = session.run(tf.contrib.memory_stats.MaxBytesInUse())
        print('mbiu', convert_size(mbiu))
        step += 1

        if train_batcher.last_batch():
            epoch += 1
            step = 0

            train_loss, train_acc, train_hits = batch_eval(session, {'x': x_train, 'y': y_train}, batch_size, epoch,
                                                           'train')
            print('train_loss', train_loss, 'train_acc', train_acc, 'train_hits', train_hits)

            vali_loss, vali_acc, vali_hits = batch_eval(session, {'x': x_vali, 'y': y_vali}, batch_size, epoch, 'vali')
            print('vali_loss', vali_loss, 'vali_acc', vali_acc, 'vali_hits', vali_hits)

            test_loss, test_acc, test_hits = batch_eval(session, {'x': x_test, 'y': y_test}, batch_size, epoch, 'test')
            print('test_loss', test_loss, 'test_acc', test_acc, 'test_hits', test_hits)

```

You can play with `skip_last`. Setting it to `False` instructs the batcher to create the last batch with size less than `batch_size` if `dataset_size % batch_size > 0`. In this case you can see the leak happening in the last batch of the validation set evaluation. Memory consuption goes suddenly from 461.72 MB to 7.41 GB. See a log here:

```
2017-11-02 21:18:28.606838: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties:
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:04:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
2017-11-02 21:18:28.606864: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080, pci bus id: 0000:04:00.0, compute capability: 6.1)
Epoch 1
epoch 1 step 1 loss 6.22924
mbiu 461.85 MB
epoch 1 step 2 loss 6.27471
mbiu 461.85 MB
epoch 1 step 3 loss 6.37816
mbiu 461.85 MB
...
epoch 1 step 55 loss 6.21585
mbiu 461.85 MB
epoch 1 step 56 loss 6.24239
mbiu 461.85 MB
epoch 1 train eval step 0
mbiu 461.85 MB
epoch 1 train eval step 1
mbiu 461.85 MB
epoch 1 train eval step 2
mbiu 461.85 MB
epoch 1 train eval step 3
mbiu 461.85 MB
...
epoch 1 train eval step 53
mbiu 461.85 MB
epoch 1 train eval step 54
mbiu 461.85 MB
epoch 1 train eval step 55
mbiu 461.85 MB
train_loss 0.048836645724 train_acc 0.00281175312808 train_hits 0.00955996063546
epoch 1 vali eval step 0
mbiu 461.85 MB
epoch 1 vali eval step 1
mbiu 461.85 MB
epoch 1 vali eval step 2
mbiu 461.85 MB
epoch 1 vali eval step 3
mbiu 461.85 MB
epoch 1 vali eval step 4
mbiu 461.85 MB
epoch 1 vali eval step 5
mbiu 461.85 MB
epoch 1 vali eval step 6
mbiu 461.85 MB
epoch 1 vali eval step 7
mbiu 7.41 GB
vali_loss 0.0538858032227 vali_acc 0.00108108108108 vali_hits 0.00216216216216
epoch 1 test eval step 0
mbiu 7.41 GB
epoch 1 test eval step 1
mbiu 7.41 GB
...
epoch 1 test eval step 13
mbiu 7.41 GB
epoch 1 test eval step 14
mbiu 7.41 GB
epoch 1 test eval step 15
mbiu 7.41 GB
test_loss 0.0507940473177 test_acc 0.00254841997961 test_hits 0.00917431192661
Epoch 2
epoch 2 step 1 loss 6.18541
mbiu 7.41 GB
epoch 2 step 2 loss 6.19003
mbiu 7.41 GB
epoch 2 step 3 loss 6.18152
mbiu 7.41 GB
...
epoch 2 test eval step 14
mbiu 7.41 GB
epoch 2 test eval step 15
mbiu 7.41 GB
test_loss 0.0509010027186 test_acc 0.00254841997961 test_hits 0.00560652395515
```

If you set `skip_last` to `True` the last betch size with a different size from the usual `batch_size` will be skipped. As you can see from this log the memory consuption stay constant at 461.72 MB from beginning to the end.

```
2017-11-02 21:14:45.784362: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties:
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:04:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
2017-11-02 21:14:45.784395: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080, pci bus id: 0000:04:00.0, compute capability: 6.1)
Epoch 1
epoch 1 step 1 loss 6.22736
mbiu 461.72 MB
epoch 1 step 2 loss 6.27873
mbiu 461.72 MB
epoch 1 step 3 loss 6.32521
mbiu 461.72 MB
epoch 1 step 4 loss 6.34492
mbiu 461.72 MB
epoch 1 step 5 loss 6.33192
mbiu 461.72 MB
epoch 1 step 6 loss 6.27755
mbiu 461.72 MB
...
epoch 2 test eval step 11
mbiu 461.78 MB
epoch 2 test eval step 12
mbiu 461.78 MB
epoch 2 test eval step 13
mbiu 461.78 MB
epoch 2 test eval step 14
mbiu 461.78 MB
test_loss 0.0475411449125 test_acc 0.00101936799185 test_hits 0.00407747196738
```

In the case of this example, the leaked memory is still within the 8 GB limit of my GPU so it doesn't go out of memory, but I have a bigger dataset and using it TF at the last batch of validation tries to allocate 33 GB and goes out of memory. To replicate that setting you should set the parameters at the beginning of the script in this way:

```
train_set_size = 2896257
vali_set_size = 97713
test_set_size = 97702
```

Hope this is enough.

**Comment by w4nderlust** [comment]
I just tested it on v1.4.0 and the problem is still there.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by Overdrivr** [comment]
I'm currently trying to run your code on a Windows 10 machine, with 2 NVidia GeForce GTX 1070.
The [memory_stats](https://github.com/tensorflow/tensorflow/issues/15412) op is not yet available on windows, so I used the following strategy to get an estimate of used VRAM, could you please comment if this approach is correct ?

I have removed the mem_stats op, and allowed growth of GPU memory allocation with

```
gpu_options = tf.GPUOptions(allow_growth=True)
with tf.Session(config=tf.ConfigProto(gpu_options=gpu_options)) as session:
```

Then using a tool like GPU-Z, I'm plotting GPU usage. Of course this gives me the memory allocated by TF, not the amount it is truly using (which should be less than that I guess).

The GPU `Memory Used` is at 352 MB at rest, after starting the script it climbs up to 1623MB (if you remove the initial 352MB, that leaves approximately ~ 1.2GB allocated by TF).

![image](https://user-images.githubusercontent.com/1294805/34879708-68927fe6-f7ae-11e7-83d7-869f988c4edb.png)

The script has not yet reached the last batch. How long does it take approximately to complete ?

**Comment by w4nderlust** [comment]
Yours is a totally different setting: different OS, different drivers, different card, different way of calculating ram, so I don't really know if that's telling anything. My best guess is that the reason for the leak i am experiencing has to do with CUDA / CuDNN for linux, so if that's true testing on Windows wouldn't help debugging unfortunately.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by jart** [comment]
Closing due to inactivity. If still a problem let us know and I'll reopen.

**Comment by w4nderlust** [comment]
Confirm that it's still an issue. Tested it with TensorFlow 1.5 with python3.5 on ubuntu 16.04 with Nvidia driver 390.12 cuda 9.1 and cuDNN 7. The leak is still there, even if now instead of taking 7G of VRAM it takes only 3.8G.

```
...
epoch 1 vali eval step 6
mbiu 472.89 MB
epoch 1 vali eval step 7
mbiu 3.8 GB
...
```

**Comment by jart** [comment]
I was hoping to see more signal boost from the community by now. But due to the seriousness of the concerns being raised here, I'd ask @zheng-xq please take another look?

**Comment by skye** [comment]
Unfortunately no one has the bandwidth to look at this right now. @w4nderlust, if you're able, paring down the script you provided to a shorter repro would be extremely helpful. I'm gonna mark contributions welcome for now since we're not actively working on it.

**Comment by w4nderlust** [comment]
Here's a much shorter version of the script, only 90 lines:

```python
import math
import numpy as np
import tensorflow as tf

# parameters
skip_last = False

sequence_length = 256
vocab_size = 20000
num_classes = 500
train_set_size = 925
epochs = 2
batch_size = 128

embedding_size = 256
filter_size = 5
num_filters = 256

# create artificial dataset
x_train = np.random.rand(train_set_size, sequence_length, embedding_size, 1)
y_train = np.random.randint(0, num_classes, [train_set_size])

# Model
input_placeholder = tf.placeholder(tf.float32, [None, sequence_length, embedding_size, 1])
label_placeholder = tf.placeholder(tf.int64, [None])
filter_shape = [filter_size, embedding_size, 1, num_filters]
weights = tf.get_variable("conv_w", filter_shape)
conv = tf.nn.conv2d(input_placeholder, weights, [1, 1, 1, 1], 'VALID')
conv = tf.nn.relu(conv)
pooled = tf.nn.max_pool(conv, [1, sequence_length - filter_size + 1, 1, 1], [1, 1, 1, 1], 'VALID')
hidden = tf.reshape(pooled, [-1, num_filters])
softmax_w = tf.get_variable("softmax_w", [num_filters, num_classes])
logits = tf.matmul(hidden, softmax_w)
probabilities = tf.nn.softmax(logits)
predictions = tf.argmax(logits, 1)
correct_predictions = tf.equal(predictions, label_placeholder)
loss = tf.losses.sparse_softmax_cross_entropy(labels=label_placeholder, logits=logits)
optimize = tf.train.AdamOptimizer(0.0025).minimize(loss)

# training
def convert_size(size_bytes):
    if size_bytes == 0:
        return "0B"
    size_name = ("B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
    i = int(math.floor(math.log(size_bytes, 1024)))
    p = math.pow(1024, i)
    s = round(size_bytes / p, 2)
    return "{} {}".format(s, size_name[i])

class Batcher(object):
    def __init__(self, x, y, batch_size, skip_last=False):
        assert len(x) == len(y)
        self.x = x
        self.y = y
        self.batch_size = batch_size
        self.total_size = len(x)
        self.skip_last = skip_last
        self.index = 0

    def next_batch(self):
        if self.last_batch():
            self.index = 0
        x_batch = self.x[self.index:self.index + self.batch_size]
        y_batch = self.y[self.index:self.index + self.batch_size]
        self.index += self.batch_size
        return x_batch, y_batch

    def last_batch(self):
        return self.index >= self.total_size or (self.skip_last and self.index + self.batch_size >= self.total_size)

with tf.Session() as session:
    tf.global_variables_initializer().run()
    epoch = 0
    step = 0
    batcher = Batcher(x_train, y_train, batch_size, skip_last)
    while epoch < epochs:
        batch_x, batch_y = batcher.next_batch()
        _, loss_val = session.run(
            [optimize, loss],
            feed_dict={input_placeholder: batch_x, label_placeholder: batch_y}
        )
        mbiu = session.run(tf.contrib.memory_stats.MaxBytesInUse())
        print('epoch: {:1d} step: {:2d} loss: {:.4f} mbiu: {}'.format(
            epoch + 1, step + 1, loss_val, convert_size(mbiu)))
        if batcher.last_batch():
            epoch += 1
            step = 0
        else:
            step += 1
```

Here is the output I'm getting from it:
```
epoch: 1 step:  1 loss: 6.2207 mbiu: 260.71 MB
epoch: 1 step:  2 loss: 6.6852 mbiu: 260.71 MB
epoch: 1 step:  3 loss: 6.5704 mbiu: 260.71 MB
epoch: 1 step:  4 loss: 6.4245 mbiu: 260.71 MB
epoch: 1 step:  5 loss: 6.2299 mbiu: 260.71 MB
epoch: 1 step:  6 loss: 6.2337 mbiu: 260.71 MB
epoch: 1 step:  7 loss: 6.2340 mbiu: 260.71 MB
epoch: 1 step:  8 loss: 6.2334 mbiu: 3.7 GB
epoch: 2 step:  1 loss: 6.2051 mbiu: 3.7 GB
epoch: 2 step:  2 loss: 6.1781 mbiu: 3.7 GB
epoch: 2 step:  3 loss: 6.1723 mbiu: 3.7 GB
epoch: 2 step:  4 loss: 6.1483 mbiu: 3.7 GB
epoch: 2 step:  5 loss: 6.1409 mbiu: 3.7 GB
epoch: 2 step:  6 loss: 6.2119 mbiu: 3.7 GB
epoch: 2 step:  7 loss: 6.2550 mbiu: 3.7 GB
epoch: 2 step:  8 loss: 6.1255 mbiu: 3.7 GB
```

As you can see the memory usage increases A LOT, more than 10x the regular usage, at the end of the epoch, when a sample with a different batch size is pushed through the graph. If I set `skip_last` to `True`, meaning the last and shorter batch is skipped, the memory usage keeps constant. This suggests there's a memory leak in `conv2d()` when dealing with batch sizes that are different than the initial one.
As this scenario is ubiquitous, it seems to me that this should be a pretty critical thing to fix rather than wait for community contributions.

**Comment by skye** [comment]
@reedwm can you take a look?

**Comment by reedwm** [comment]
Here is a simpler example the reproduces the problem

```python
import numpy as np
import tensorflow as tf

sequence_length = 256
embedding_size = 256
filter_size = 5
num_filters = 256

# Model
input_placeholder = tf.placeholder(tf.float32, [None, sequence_length, embedding_size, 1])
label_placeholder = tf.placeholder(tf.int64, [None])
filter_shape = [filter_size, embedding_size, 1, num_filters]
weights = tf.get_variable("conv_w", filter_shape)
conv = tf.nn.conv2d(input_placeholder, weights, [1, 1, 1, 1], 'VALID')

with tf.Session() as session:
  tf.global_variables_initializer().run()
  sizes = [128] * 2 + [29] + [128] * 2
  for size in sizes:
    batch_x = np.zeros([size, sequence_length, embedding_size, 1])
    batch_y = np.zeros([size])
    loss_val = session.run(
        conv.op,
        feed_dict={input_placeholder: batch_x, label_placeholder: batch_y}
    )
    mbiu, biu = session.run([tf.contrib.memory_stats.MaxBytesInUse(),
                             tf.contrib.memory_stats.BytesInUse()])
    print('mbiu: {} MB, bui: {} MB'.format(mbiu / 2**20, biu / 2**20))
```

The output is:

```
mbiu: 255 MB, bui: 1 MB
mbiu: 255 MB, bui: 1 MB
mbiu: 3766 MB, bui: 1 MB
mbiu: 3766 MB, bui: 1 MB
mbiu: 3766 MB, bui: 1 MB
```

This is no memory leak, because the current bytes-in-use always goes back to 1MB after the session finishes running. But it is weird how running a single step with a batch-size of 29 causes future steps with a batch size of 128 to use more memory.

@yzhwang can you take a look? My post has all the info of the root problem, so I don't think you need to read the previous posts. Is this an issue with autotune?

**Comment by reedwm** [comment]
Actually I just realized that MaxBytesInUse stays at 3766 MB, since it is never reset, so once it reaches 3766 MB it will never go down.

So, it seems the issue is that a batch size of 29 uses more memory than a batch size of 128. My guess is that with a batch size of 29, the faster algorithm is different than with a batch size of 128. The batch-size-29 algorithm probably uses a lot more memory than the batch-size-128 algorithm, which explains the difference

@yzhwang can you confirm?

**Comment by yzhwang** [comment]
Yes, I tested with TensorFlow 1.5.0. By turning off autotune with this code:
`os.environ["TF_CUDNN_USE_AUTOTUNE"] = "0"`
peak memory drops to 97.5MB, otherwise I get the same result as @reedwm .

The reason for this is that currently, when we use autotune to pick the best algorithm, we use running time as the only metrics. We should start considering memory consumption too. This is on my todo list, and I will update the autotune metrics but it won't happen within weeks.

**Comment by reedwm** [comment]
@yzhwang, thanks for the info. Closing this issue, since considering memory consumption is a separate issue.

**Comment by w4nderlust** [comment]
Thank you guys for shedding light on the issue. In my specific case, this additional usage of memory for a smaller batch size was making the model go out of memory, not immediately but at the end of the epoch. So if I may have a suggestion. Now I know that setting TF_CUDNN_USE_AUTOTUNE I can keep the memory usage constant.
So may I suggest adding an option to ConfigProto? Setting it the user would know that he can rely on constant memory consumption and he can tune the size of the model and the batch size accordingly, without weird surprises at the end of the epoch that will result in out of memory errors.

**Comment by zheng-xq** [comment]
I am suspecting something weird is happening. With autotuning, it is not supposed to use much more memory than if autotune is disabled

1. During the measurement phase, if an algorithm failed to allocate memory, it is not a fatal error. TF just removes that algorithm from candidacy.
2. During the execution phase, even if the op uses scratch memory, it almost immediately releases for later. Only concurrent kernels might be impacted by this. But even with that, other ops should retry and as if no extra memory was used.

Something might not be following this protocol. Disabling autotune has major performance impact. Also it should have too much memory overhead. So let's root cause this first. 

Reopening the issue and ask @yzhwang to take a look.

**Comment by yzhwang** [comment]
@w4nderlust We had a discussion offline. So as @zheng-xq just mentioned, autotune could use a large chunk of memory in two cases. In the measurement phase, I think he has explained it very well why there shouldn't be an OOM or any other failure. Just to provide you more info as to help you understand what happens in the execution phase:
There will be two algorithms recorded in autotune: the primary one which uses scratch memory, a secondary one which uses zero scratch memory, and regarding cuDNN's documentation, as long as the input shape and other parameters are legal, there should always be a default algorithm without scratch memory working fine.
Suppose autotune find a best performing algorithm to use but in the execution phase it fails to allocate the scratch memory, we get a warning, but not a fatal error:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2400
Then in the following few lines, we will fall back to that default algorithm that uses no scratch memory. If an error happens there, it means one of two things:
1) there is a cuDNN bug, since the default algorithm doesn't work for a legal set of parameters;
2) we somehow fail to set the default algorithm for a legal set of parameters that should have a default algorithm.
In either case, we would be happy to fix that for you.

My suggestion for you is to try more batch size for validation, and see if there will be any error happening during the execution phase of conv. Note that if an OOM happens before the execution, it is the intended behavior as it would imply that TF doesn't have enough memory to even hold the input tensor. If you see error during the execution phase of conv, please report back and we are committed to fix that for you.

Also note that MaxBytesInUse() is a watermark. So it doesn't reflect the current memory in use, it only reflects the largest memory in use ever. So it could be a large one as long as one of the internal algorithm needs a large chunk of scratch memory.

**Comment by w4nderlust** [comment]
@yzhwang thank you really much for your detailed explanation and for taking the time to write it. I will test if the OOM that I was getting originally still happens now (with updated TF and CUDA/cuDNN) and in case will provide some code to replicate it. Thank you again.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 95 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 111 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by poxvoculi** [comment]
Closing due to inactivity.

**Comment by ChunCheong** [comment]
I am facing the same issue and somehow google has brought me here...
I saw OOM error when trying to apply Conv2D even the input and output size is well below the ram i have on GPU.
I also see setting autotune has reduced the peak memory by a bit.

Any more recent advice? Many Thanks!

Some info:
tensorflow: 1.14 
GPU:Titan XP 
cuda: 9.0

**Comment by digital10111** [comment]
> I am facing the same issue and somehow google has brought me here...
> I saw OOM error when trying to apply Conv2D even the input and output size is well below the ram i have on GPU.
> I also see setting autotune has reduced the peak memory by a bit.
> 
> Any more recent advice? Many Thanks!
> 
> Some info:
> tensorflow: 1.14
> GPU:Titan XP
> cuda: 9.0

I saw the same issue today with tf2.0. Before the conv2d operator the usage was like 0.3gb and right after I get the result from the operator, the gpu memory usage 4.8gb constantly.
